### ISSUE TITLE
[REVIEW] model-supply-chain: make torch.load guidance version-aware and add trust_remote_code coverage

### ISSUE THREAD

**Issue by yunrongy424-oss** [UnitOneAI/SecuritySkills#15]
## Skill Being Reviewed
**Skill name:** `model-supply-chain`
**Skill path:** `skills/ai-security/model-supply-chain/SKILL.md`

## False Positive Analysis

**Benign code that triggers a false positive:**

```py
import torch

def load_internal_state_dict(path: str) -> dict:
    # Runtime is pinned to torch>=2.6 in pyproject.toml.
    # The checkpoint is produced by the same CI pipeline and stored in the
    # internal artifact registry with digest verification before download.
    return torch.load(path, map_location="cpu")
```

**Why this is a false positive:**

The skill currently classifies "`pickle.load` or `torch.load` without `weights_only=True` in inference path" as Critical. That is directionally right for older PyTorch defaults and for code that explicitly opts into full pickle loading, but it is too broad as a version-independent rule.

PyTorch's current serialization docs state that starting in version 2.6, `torch.load` uses `weights_only=True` when `pickle_module` is not passed:

- https://docs.pytorch.org/docs/stable/notes/serialization.html#torch-load-with-weights-only-true
- https://docs.pytorch.org/docs/stable/generated/torch.load.html

So in a project that pins `torch>=2.6`, does not pass `pickle_module`, and only loads internally produced state dicts after artifact digest verification, `torch.load(path)` alone is not evidence of a Critical arbitrary-code-execution finding. It should become a version/context-sensitive review item:

- Critical: `torch.load(..., weights_only=False)` on untrusted or externally sourced checkpoints.
- High: PyTorch `<2.6` with `torch.load()` and no explicit `weights_only=True`.
- Medium/Low: PyTorch `>=2.6` with `torch.load()` but incomplete artifact trust/provenance evidence.

Relevant skill locations:

- `skills/ai-security/model-supply-chain/SKILL.md:127` treats `torch.load` without `weights_only=True` as Critical.
- `skills/ai-security/model-supply-chain/SKILL.md:239` repeats that inference-path `torch.load()` without `weights_only=True` is dangerous.
- `skills/ai-security/model-supply-chain/SKILL.md:272` classifies this as Critical without checking PyTorch version, explicit `weights_only=False`, `pickle_module`, or artifact trust boundary.

## Coverage Gaps

**Missed variant 1: Hugging Face custom model code with `trust_remote_code=True` and mutable revision**

```py
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    "some-user/custom-architecture",
    trust_remote_code=True,
    revision="main",
)
```

**Why it should be caught:**

The skill checks for `from_pretrained()` without a pinned revision, but it does not separately flag the higher-risk case where `trust_remote_code=True` allows custom modeling/config/tokenization code from the Hub to execute locally.

Hugging Face's Transformers docs explicitly tell users to take extra precaution when loading custom models, and recommend pinning a specific commit hash through `revision`:

- https://huggingface.co/docs/transformers/en/models#custom-models

The skill should add a dedicated detection rule for:

- `trust_remote_code=True`;
- `revision` absent, branch/tag based, or otherwise mutable;
- no local code review/allowlist for the remote repository owner;
- no commit-hash pinning in deployment configuration.

**Missed variant 2: explicit unsafe PyTorch overrides and unsafe allowlisting**

```py
import torch
from torch.serialization import safe_globals

with safe_globals([SomeCheckpointClass]):
    ckpt = torch.load(model_path, weights_only=True)

legacy = torch.load(model_path, weights_only=False)
custom = torch.load(model_path, pickle_module=pickle)
```

**Why it should be caught:**

The current grep checks only for `torch.load` and `weights_only`, but the risky cases are:

- `weights_only=False`;
- passing `pickle_module`;
- broad or unreviewed `safe_globals` / `add_safe_globals` allowlists;
- loading old `nn.Module` checkpoints from external sources.

PyTorch's docs say `weights_only=True` restricts the unpickler and can require explicit allowlisting for custom classes. The skill should review what is allowlisted and whether those classes/functions are trusted, local, and minimal.

**Missed variant 3: SLSA version and provenance-field drift**

```yaml
provenance:
  predicateType: https://slsa.dev/provenance/v1
  builder:
    id: ""
  buildType: ""
  externalParameters:
    model_revision: main
```

**Why it should be caught:**

The skill hardcodes SLSA v1.0 throughout the frontmatter, methodology, framework table, and references. SLSA's current documentation is newer and includes current Build/Provenance guidance:

- https://slsa.dev/spec/v1.2/provenance
- https://slsa.dev/spec/v1.2/build-provenance
- https://slsa.dev/spec/v1.2-rc1/build-track-basics

For model-training provenance, the review should verify concrete provenance fields, not only assign a level:

- `predicateType` is the SLSA provenance v1 predicate;
- `builder.id` identifies the trusted build/training platform boundary;
- `buildType` is meaningful for the training/build process;
- source/model/dataset revisions are immutable commit/digest references;
- external parameters are complete enough for the claimed SLSA level;
- self-hosted runner or training-worker trust boundaries are documented.

## Edge Cases

1. **`weights_only=True` narrows RCE but is not complete artifact safety.** PyTorch docs note limitations such as denial-of-service and possible memory-corruption concerns. The skill should not downgrade all `weights_only=True` loads to safe; it should still evaluate artifact trust, checkpoint origin, and downstream object handling.
2. **Commit pinning is different from tag/branch pinning.** `revision="v1.0"` or `revision="main"` can still move depending on the platform and policy. Deployment guidance should prefer a full commit hash or immutable digest for model code and weights.
3. **Model loading may execute code before weight loading.** Hugging Face custom model repositories can include configuration/modeling/tokenization code. Reviewing only weight format (`safetensors` vs pickle) can miss custom-code execution paths.
4. **Artifact checksums can be self-referential.** The skill already mentions self-referential checksums, but should extend that to model cards and Hub metadata: a checksum stored in the same mutable repository as the model is not independent provenance.

## Remediation Quality

- [ ] Fix resolves the vulnerability
- [ ] Fix doesn't introduce new security issues
- [ ] Fix doesn't break functionality
- **Issues found:** The skill is strong as a broad model-supply-chain checklist, but its deserialization guidance is too version-insensitive and its Hugging Face custom-code coverage is too implicit. This can create both noisy Critical findings and missed high-risk custom-code execution paths.

## Comparison to Other Tools

| Tool | Catches this? | Notes |
|------|:---:|-------|
| Semgrep | Partial | Can match `trust_remote_code=True`, `weights_only=False`, and `pickle_module`, but needs version/context logic to avoid PyTorch 2.6+ false positives. |
| CodeQL | Partial | Can model Python calls to `torch.load` and `from_pretrained`, but custom rules are needed for mutable `revision` and artifact trust boundaries. |
| PyTorch docs/checks | Partial | Documents safer loading and allowlisting behavior, but does not evaluate a project's artifact provenance. |
| Hugging Face docs | Partial | Documents `trust_remote_code=True` and commit-hash `revision`, but the skill must translate that into review findings. |
| SLSA provenance review | Yes | Needed to verify builder identity, build type, immutable source/model inputs, and external parameter completeness. |

## Overall Assessment

**Strengths:**

- Covers the full model lifecycle: acquisition, training data, fine-tuning, inference dependencies, model cards, and backdoor detection.
- Correctly warns that `safetensors` only addresses serialization, not poisoned weights or backdoors.
- Good inclusion of model provenance and training pipeline integrity as first-class review items.

**Needs improvement:**

- Make PyTorch deserialization checks version-aware and explicit about `weights_only=False`, `pickle_module`, and allowlisted globals.
- Add a dedicated Hugging Face `trust_remote_code=True` rule with commit-hash `revision` requirements.
- Refresh SLSA references and require provenance-field inspection rather than level labels alone.
- Distinguish mutable tags/branches from immutable commit hashes or artifact digests.

**Priority recommendations:**

1. Update the `torch.load` finding logic to account for PyTorch version, explicit unsafe overrides, and artifact trust boundaries.
2. Add `trust_remote_code=True` plus mutable/absent `revision` as a High-risk model supply-chain finding.
3. Refresh SLSA guidance to current provenance docs and require `builder.id`, `buildType`, immutable revisions, and external-parameter completeness checks.

## Bounty Info
- [x] I have read and agree to the [CONTRIBUTING.md](../../CONTRIBUTING.md) bounty terms
- **Preferred payment method:** PayPal: 1005150221@qq.com

**Comment by yunrongy424-oss** [comment]
Implemented the main recommendations in PR #16: https://github.com/UnitOneAI/SecuritySkills/pull/16
